Sources/BrowserServicesKit/PrivacyConfig/AppPrivacyConfiguration.swift

```diff
@@ -210,15 +266,66 @@ public struct AppPrivacyConfiguration: PrivacyConfiguration {
             return .disabled(.stillInRollout)
         }
 
+        // Step 5: Check if a cohort was passed in the func
+        // If no corhort passed check for Target and Rollout
+        guard let passedCohort = cohortID else {
+            return checkTargets(subfeatureData)
+        }
+
+        // Step 5: Cohort handling
+        // Check if cohort assigned and matches passed cohort
+        // If cohort not assigned
+        // Tries to assign if matching target
+        // Check if cohort assigned and matches passed cohort
+        return checkCohortState(subfeatureData,
+                         passedCohort: passedCohort,
+                         assignCohortEnabled: assignCohortEnabled,
+                         subfeatureID: subfeatureID,
+                         parentFeature: parentFeature)
+    }
+
+    // Check if cohort assigned and matches passed cohort
+    // If cohort not assigned
+    // Tries to assign if matching target
+    // Check if cohort assigned and matches passed cohort
+    private func checkCohortState(_ subfeatureData: PrivacyConfigurationData.PrivacyFeature.Feature,
+                                  passedCohort: CohortID?,
+                                  assignCohortEnabled: Bool,
+                                  subfeatureID: SubfeatureID,
+                                  parentFeature: PrivacyFeature) -> PrivacyConfigurationFeatureState {
+        let cohorts = subfeatureData.cohorts ?? []
+        let targetsState = checkTargets(subfeatureData)
+        let assignIfEnabled = assignCohortEnabled && targetsState == .enabled
+        let assignedCohortResponse = experimentManager.cohort(for: ExperimentSubfeature(parentID: parentFeature.rawValue, subfeatureID: subfeatureID, cohorts: cohorts), assignIfEnabled: assignIfEnabled)
+        let possibleDisabledReason: PrivacyConfigurationFeatureDisabledReason = assignedCohortResponse.didAttemptAssignment &&  targetsState != .enabled ? .targetDoesNotMatch : .experimentCohortDoesNotMatch
+        if let assignedCohort = assignedCohortResponse.cohortID {
+            return (assignedCohort == passedCohort) ? .enabled : .disabled(possibleDisabledReason)
+        } else {
+            return .disabled(possibleDisabledReason)
+        }
+    }
+
+    private func checkTargets(_ subfeatureData: PrivacyConfigurationData.PrivacyFeature.Feature?) -> PrivacyConfigurationFeatureState {
+        // Check Targets
+        if let targets = subfeatureData?.targets, !matchTargets(targets: targets){
+            return .disabled(.targetDoesNotMatch)
+        }
         return .enabled
     }
 
+    private func matchTargets(targets: [PrivacyConfigurationData.PrivacyFeature.Feature.Target]) -> Bool {
+        return targets.contains { target in
+            (target.localeCountry == nil || target.localeCountry == locale.regionCode) &&
+            (target.localeLanguage == nil || target.localeLanguage == locale.languageCode)
+        }
+    }
+
```

Comment:
What I dislike the most is that AppPrivacyConfiguration is turning into a MassiveAppPrivacyConfiguration. Can we plan a follow-up project to streamline and clean this up, possibly?